File: experiments/engine-kt/plan/src/main/kotlin/com/garive/eng/kt/plan/PlanLifecycle.kt
package com.garive.eng.kt.plan

/** Closed lifecycle for one immutable Plan revision. */
public enum class PlanState(public val wireName: String) {
    /** Valid proposal awaiting Runtime authority. */ PROPOSED("proposed"),
    /** Authoritative revision with no started step yet. */ ADOPTED("adopted"),
    /** At least one step started or remains available. */ RUNNING("running"),
    /** Plan-level continuation is required. */ SUSPENDED("suspended"),
    /** Every step and criterion reduction was verified. */ COMPLETED("completed"),
    /** Explicit terminal failure. */ FAILED("failed"),
    /** A newer revision atomically replaced this revision. */ SUPERSEDED("superseded"),
    /** Runtime authority rejected this proposal. */ REJECTED("rejected"),
}

/** Closed progress state for one declared step. */
public enum class StepState(public val wireName: String) {
    /** Dependencies are not complete. */ PENDING("pending"),
    /** Dependencies and bounds admit a claim. */ READY("ready"),
    /** Fenced claim exists, no attempt started. */ CLAIMED("claimed"),
    /** One attempt owns a Kernel Execution. */ RUNNING("running"),
    /** Typed continuation is required. */ SUSPENDED("suspended"),
    /** Verified terminal evidence exists. */ COMPLETED("completed"),
    /** Last attempt failed; policy may admit retry. */ FAILED("failed"),
}

/** Immutable public progress for one step. */
public data class StepProgress(public val state: StepState, public val attempts: Int)

/** One requested pure Plan/step transition after Runtime validation. */
public sealed interface PlanTransition {
    /** Adopt a valid proposal. */ public data object Adopt : PlanTransition
    /** Adopt with Runtime-verified completed steps from a prior revision. */
    public data class AdoptWithCarryForward(public val stepIds: Set<PlanStepId>) : PlanTransition
    /** Reject a proposal. */ public data object Reject : PlanTransition
    /** Suspend Plan-level dispatch. */ public data object Suspend : PlanTransition
    /** Resume Plan-level dispatch. */ public data object Resume : PlanTransition
    /** Supersede with another validated revision. */ public data object Supersede : PlanTransition
    /** Explicitly terminalize as failed. */ public data object Fail : PlanTransition
    /** Complete only with full step and criterion evidence. */
    public data class Complete(public val criteriaComplete: Boolean) : PlanTransition
    /** Claim one Ready step. */ public data class Claim(public val stepId: PlanStepId) : PlanTransition
    /** Expire a never-started claim. */ public data class ExpireClaim(public val stepId: PlanStepId) : PlanTransition
    /** Start one claimed attempt. */ public data class Start(public val stepId: PlanStepId) : PlanTransition
    /** Complete one running attempt. */ public data class CompleteStep(public val stepId: PlanStepId) : PlanTransition
    /** Suspend one running attempt. */ public data class SuspendStep(public val stepId: PlanStepId) : PlanTransition
    /** Resume after continuation resolution. */ public data class ResumeStep(public val stepId: PlanStepId) : PlanTransition
    /** Fail one running attempt. */ public data class FailStep(public val stepId: PlanStepId) : PlanTransition
    /** Admit a bounded retry. */ public data class RetryStep(public val stepId: PlanStepId) : PlanTransition
}

/** Immutable Plan projection after one contiguous transition prefix. */
public class PlanSnapshot private constructor(
    public val definition: PlanDefinitionV1,
    public val state: PlanState,
    progress: Map<PlanStepId, StepProgress>,
    public val totalAttempts: Int,
) {
    private val progress: Map<PlanStepId, StepProgress> = progress.toMap()

    /** Returns progress for one declared step. */
    public fun step(stepId: PlanStepId): StepProgress? = progress[stepId]

    /** Returns Ready steps in declaration order. */
    public fun readySteps(): List<PlanStepId> = definition.steps
        .filter { progress.getValue(it.stepId).state == StepState.READY }
        .map(PlanStepV1::stepId)

    /** Applies one legal transition without allocating Runtime identities. */
    public fun apply(transition: PlanTransition): PlanResult<PlanSnapshot> {
        if (state in TERMINAL) return failure(PlanErrorCode.PLAN_TRANSITION_INVALID)
        return when (transition) {
            PlanTransition.Adopt -> if (state == PlanState.PROPOSED) updated(PlanState.ADOPTED, progress).refresh()
            else invalid()
            is PlanTransition.AdoptWithCarryForward -> carryForward(transition.stepIds)
            PlanTransition.Reject -> if (state == PlanState.PROPOSED) success(updated(PlanState.REJECTED, progress)) else invalid()
            PlanTransition.Suspend -> if (state == PlanState.RUNNING) success(updated(PlanState.SUSPENDED, progress)) else invalid()
            PlanTransition.Resume -> if (state == PlanState.SUSPENDED) updated(PlanState.RUNNING, progress).refresh()
            else invalid()
            PlanTransition.Supersede -> if (state in ACTIVE) success(updated(PlanState.SUPERSEDED, progress)) else invalid()
            PlanTransition.Fail -> if (state == PlanState.RUNNING || state == PlanState.SUSPENDED) {
                success(updated(PlanState.FAILED, progress))
            } else invalid()
            is PlanTransition.Complete -> if (state == PlanState.RUNNING && transition.criteriaComplete &&
                progress.values.all { it.state == StepState.COMPLETED }
            ) success(updated(PlanState.COMPLETED, progress)) else invalid()
            is PlanTransition.Claim -> claim(transition.stepId)
            is PlanTransition.ExpireClaim -> replace(transition.stepId, StepState.CLAIMED, StepState.READY)
            is PlanTransition.Start -> start(transition.stepId)
            is PlanTransition.CompleteStep -> replace(
                transition.stepId, StepState.RUNNING, StepState.COMPLETED, refresh = true,
            )
            is PlanTransition.SuspendStep -> replace(transition.stepId, StepState.RUNNING, StepState.SUSPENDED)
            is PlanTransition.ResumeStep -> replace(
                transition.stepId, StepState.SUSPENDED, StepState.PENDING, refresh = true,
            )
            is PlanTransition.FailStep -> replace(transition.stepId, StepState.RUNNING, StepState.FAILED)
            is PlanTransition.RetryStep -> retry(transition.stepId)
        }
    }

    private fun claim(stepId: PlanStepId): PlanResult<PlanSnapshot> {
        val active = progress.values.count { it.state == StepState.CLAIMED || it.state == StepState.RUNNING }
        if (state !in setOf(PlanState.ADOPTED, PlanState.RUNNING) ||
            active >= definition.bounds.maxParallelReady || progress[stepId]?.state != StepState.READY
        ) return failure(PlanErrorCode.STEP_NOT_READY)
        return replace(stepId, StepState.READY, StepState.CLAIMED)
    }

    private fun carryForward(stepIds: Set<PlanStepId>): PlanResult<PlanSnapshot> {
        if (state != PlanState.PROPOSED || stepIds.any { id ->
                definition.step(id)?.dependsOn?.all(stepIds::contains) != true
            }
        ) return invalid()
        val next = progress.mapValues { (id, value) ->
            if (id in stepIds) value.copy(state = StepState.COMPLETED) else value
        }
        val nextState = if (stepIds.isEmpty()) PlanState.ADOPTED else PlanState.RUNNING
        return updated(nextState, next).refresh()
    }

    private fun start(stepId: PlanStepId): PlanResult<PlanSnapshot> {
        val current = progress[stepId] ?: return invalid()
        val limit = definition.step(stepId)?.maxAttempts ?: return invalid()
        if (current.state != StepState.CLAIMED) return invalid()
        if (current.attempts >= limit || totalAttempts >= definition.bounds.maxTotalAttempts) {
            return failure(PlanErrorCode.PLAN_BOUND_EXCEEDED)
        }
        val next = progress + (stepId to StepProgress(StepState.RUNNING, current.attempts + 1))
        return PlanResult.Success(PlanSnapshot(definition, PlanState.RUNNING, next, totalAttempts + 1))
    }

    private fun retry(stepId: PlanStepId): PlanResult<PlanSnapshot> {
        val current = progress[stepId] ?: return invalid()
        val limit = definition.step(stepId)?.maxAttempts ?: return invalid()
        if (current.state != StepState.FAILED) return invalid()
        if (current.attempts >= limit || totalAttempts >= definition.bounds.maxTotalAttempts) {
            return failure(PlanErrorCode.PLAN_BOUND_EXCEEDED)
        }
        return replace(stepId, StepState.FAILED, StepState.PENDING, refresh = true)
    }

    private fun replace(
        stepId: PlanStepId,
        expected: StepState,
        next: StepState,
        refresh: Boolean = false,
    ): PlanResult<PlanSnapshot> {
        val current = progress[stepId] ?: return invalid()
        if (current.state != expected) return invalid()
        val snapshot = PlanSnapshot(
            definition, state, progress + (stepId to current.copy(state = next)), totalAttempts,
        )
        return if (refresh) snapshot.refresh() else PlanResult.Success(snapshot)
    }

    private fun refresh(): PlanResult<PlanSnapshot> {
        if (state !in setOf(PlanState.ADOPTED, PlanState.RUNNING) ||
            totalAttempts >= definition.bounds.maxTotalAttempts
        ) return PlanResult.Success(this)
        val completed = progress.filterValues { it.state == StepState.COMPLETED }.keys
        val next = progress.toMutableMap()
        definition.steps.forEach { step ->
            val current = next.getValue(step.stepId)
            if (current.state == StepState.PENDING && current.attempts < step.maxAttempts &&
                step.dependsOn.all(completed::contains)
            ) next[step.stepId] = current.copy(state = StepState.READY)
        }
        return PlanResult.Success(PlanSnapshot(definition, state, next, totalAttempts))
    }

    private fun updated(next: PlanState, values: Map<PlanStepId, StepProgress>): PlanSnapshot =
        PlanSnapshot(definition, next, values, totalAttempts)

    private fun success(value: PlanSnapshot): PlanResult<PlanSnapshot> = PlanResult.Success(value)

    private fun invalid(): PlanResult<PlanSnapshot> = failure(PlanErrorCode.PLAN_TRANSITION_INVALID)

    public companion object {
        private val TERMINAL = setOf(
            PlanState.COMPLETED, PlanState.FAILED, PlanState.SUPERSEDED, PlanState.REJECTED,
        )
        private val ACTIVE = setOf(PlanState.ADOPTED, PlanState.RUNNING, PlanState.SUSPENDED)

        /** Creates a Proposed projection with every step Pending. */
        public fun create(definition: PlanDefinitionV1): PlanSnapshot = PlanSnapshot(
            definition,
            PlanState.PROPOSED,
            definition.steps.associate { it.stepId to StepProgress(StepState.PENDING, 0) },
            0,
        )
    }
}

File: engine/plan/tests/plan_lifecycle.rs
use std::collections::BTreeSet;

use garive_plan::{
    PlanBoundsV1, PlanCapabilityReference, PlanDefinitionV1, PlanErrorCode, PlanId, PlanSnapshot,
    PlanState, PlanStepId, PlanStepV1, PlanTransition, StepState,
};

#[test]
fn adopted_dag_unlocks_in_declaration_order_and_completes_explicitly() {
    let prepare = id("prepare");
    let deliver = id("deliver");
    let adopted = PlanSnapshot::new(definition(2, 4, 2))
        .apply(PlanTransition::Adopt)
        .unwrap();
    assert_eq!(adopted.ready_steps(), vec![&prepare]);

    let running = adopted
        .apply(PlanTransition::Claim(prepare.clone()))
        .unwrap()
        .apply(PlanTransition::Start(prepare.clone()))
        .unwrap();
    assert_eq!(running.state(), PlanState::Running);
    assert_eq!(running.step(&prepare).unwrap().attempts(), 1);
    let progressed = running
        .apply(PlanTransition::CompleteStep(prepare))
        .unwrap();
    assert_eq!(progressed.ready_steps(), vec![&deliver]);
    let finished_steps = progressed
        .apply(PlanTransition::Claim(deliver.clone()))
        .unwrap()
        .apply(PlanTransition::Start(deliver.clone()))
        .unwrap()
        .apply(PlanTransition::CompleteStep(deliver))
        .unwrap();
    assert_eq!(
        finished_steps
            .apply(PlanTransition::Complete {
                criteria_complete: false,
            })
            .unwrap_err()
            .code(),
        PlanErrorCode::PlanTransitionInvalid
    );
    let completed = finished_steps
        .apply(PlanTransition::Complete {
            criteria_complete: true,
        })
        .unwrap();
    assert_eq!(completed.state(), PlanState::Completed);
}

#[test]
fn claim_capacity_and_attempt_limits_fail_closed() {
    let first = id("first");
    let second = id("second");
    let adopted = PlanSnapshot::new(independent_definition(1, 1))
        .apply(PlanTransition::Adopt)
        .unwrap();
    assert_eq!(adopted.ready_steps(), vec![&first, &second]);
    let claimed = adopted.apply(PlanTransition::Claim(first.clone())).unwrap();
    assert_eq!(
        claimed
            .apply(PlanTransition::Claim(second))
            .unwrap_err()
            .code(),
        PlanErrorCode::StepNotReady
    );
    let failed = claimed
        .apply(PlanTransition::Start(first.clone()))
        .unwrap()
        .apply(PlanTransition::FailStep(first.clone()))
        .unwrap();
    assert_eq!(failed.step(&first).unwrap().state(), StepState::Failed);
    assert_eq!(
        failed
            .apply(PlanTransition::RetryStep(first))
            .unwrap_err()
            .code(),
        PlanErrorCode::PlanBoundExceeded
    );
}

#[test]
fn never_started_claim_can_expire_but_started_attempt_cannot_replay() {
    let prepare = id("prepare");
    let claimed = PlanSnapshot::new(definition(2, 4, 2))
        .apply(PlanTransition::Adopt)
        .unwrap()
        .apply(PlanTransition::Claim(prepare.clone()))
        .unwrap();
    let ready = claimed
        .apply(PlanTransition::ExpireClaim(prepare.clone()))
        .unwrap();
    assert_eq!(ready.step(&prepare).unwrap().state(), StepState::Ready);
    let running = ready
        .apply(PlanTransition::Claim(prepare.clone()))
        .unwrap()
        .apply(PlanTransition::Start(prepare.clone()))
        .unwrap();
    assert_eq!(
        running
            .apply(PlanTransition::ExpireClaim(prepare))
            .unwrap_err()
            .code(),
        PlanErrorCode::PlanTransitionInvalid
    );
}

#[test]
fn carry_forward_adopts_only_dependency_closed_completed_steps() {
    let prepare = id("prepare");
    let deliver = id("deliver");
    let carried = PlanSnapshot::new(definition(2, 4, 2))
        .apply(PlanTransition::AdoptWithCarryForward(BTreeSet::from([
            prepare.clone(),
        ])))
        .unwrap();
    assert_eq!(carried.state(), PlanState::Running);
    assert_eq!(
        carried.step(&prepare).unwrap().state(),
        StepState::Completed
    );
    assert_eq!(carried.step(&prepare).unwrap().attempts(), 0);
    assert_eq!(carried.ready_steps(), vec![&deliver]);
    assert_eq!(
        PlanSnapshot::new(definition(2, 4, 2))
            .apply(PlanTransition::AdoptWithCarryForward(BTreeSet::from([
                deliver,
            ])))
            .unwrap_err()
            .code(),
        PlanErrorCode::PlanTransitionInvalid
    );
}

fn definition(parallel: u32, total_attempts: u32, step_attempts: u32) -> PlanDefinitionV1 {
    build(
        vec![
            step("prepare", [], ["accepted"], step_attempts),
            step("deliver", ["prepare"], ["artifact"], step_attempts),
        ],
        parallel,
        total_attempts,
    )
}

fn independent_definition(parallel: u32, step_attempts: u32) -> PlanDefinitionV1 {
    build(
        vec![
            step("first", [], ["accepted"], step_attempts),
            step("second", [], ["artifact"], step_attempts),
        ],
        parallel,
        4,
    )
}

fn build(steps: Vec<PlanStepV1>, parallel: u32, total_attempts: u32) -> PlanDefinitionV1 {
    let capabilities = BTreeSet::from([capability()]);
    PlanDefinitionV1::new(
        PlanId::new("plan-1").unwrap(),
        1,
        "goal-1",
        2,
        digest('a'),
        digest('b'),
        digest('c'),
        "safety-v1",
        steps,
        PlanBoundsV1::new(4, parallel, total_attempts, None, None).unwrap(),
        &set(["accepted", "artifact"]),
        &BTreeSet::new(),
        &capabilities,
    )
    .unwrap()
}

fn step<const D: usize, const C: usize>(
    value: &str,
    dependencies: [&str; D],
    criteria: [&str; C],
    max_attempts: u32,
) -> PlanStepV1 {
    PlanStepV1::new(
        id(value),
        format!("Complete {value}"),
        dependencies.into_iter().map(id),
        criteria.into_iter().map(str::to_owned),
        [capability()],
        [digest('d')],
        max_attempts,
    )
    .unwrap()
}

fn id(value: &str) -> PlanStepId {
    PlanStepId::new(value).unwrap()
}

fn capability() -> PlanCapabilityReference {
    PlanCapabilityReference::new("tools", "catalogue-v1").unwrap()
}

fn set<const N: usize>(values: [&str; N]) -> BTreeSet<String> {
    values.into_iter().map(str::to_owned).collect()
}

fn digest(character: char) -> String {
    character.to_string().repeat(64)
}
